pkg/cmd/server/crypto/crypto.go

```diff
@@ -351,6 +337,16 @@ func NewSerialFileGenerator(serialFile string, createIfNeeded bool) (*SerialFile
 func (s *SerialFileGenerator) Next(template *x509.Certificate) (int64, error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
+
+	// do a best effort check to make sure concurrent external writes are not occurring to the underlying serial file
+	serial, err := fileToSerial(s.SerialFile)
```

Comment:
This seems futile… we know this utility isn't safe to run concurrently. Making the occurrences of races harder to notice doesn't really make it actually safe. Randomizing the serial would be better if we want to allow concurrent running

Reply:
Is there any way to lock the serial while the utility runs ?

Reply:
> Making the occurrences of races harder to notice doesn't really make it actually safe.

How does this do that?  All it does is blow up when it definitely detects that something is writing to the same file.  Thus, instead of always silently ignoring and overwriting changes to the underlying file, it has some chance of noticing so the user is clearly aware that they are doing something wrong.  If it fails to detect concurrent writes, it is still no worse than not checking at all.

> Randomizing the serial would be better if we want to allow concurrent running

I do not intend on allowing concurrent running - I just want to fail hard if I notice concurrent running.

> Is there any way to lock the serial while the utility runs ?

I am unaware of something like that.  Would not that be operation system / file system specific?

Reply:
It depends on how locking is done. A traditional crude way of doing portable whole-file locking is to create a lock file. The problem with a lockfile is the need to manually clean it up if the tool that creates it dies before proper cleanup, so it is not a great tool.
Given here you are just detecting IMO your approach is fine as it is.